// @microsoft/rush-lib/lib/cli/logic/test/Telemetry.test.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const RushConfiguration_1 = require("../../../data/RushConfiguration");
const Rush_1 = require("../../../Rush");
const Telemetry_1 = require("../Telemetry");
describe('Telemetry', () => {
    it('adds data to store if telemetry is enabled', () => {
        const filename = path.resolve(path.join(__dirname, './telemetry/telemetryEnabled.json'));
        const rushConfig = RushConfiguration_1.RushConfiguration.loadFromConfigurationFile(filename);
        const telemetry = new Telemetry_1.Telemetry(rushConfig);
        const logData1 = {
            name: 'testData1',
            duration: 100,
            result: 'Succeeded',
            timestamp: new Date().getTime(),
            platform: process.platform,
            rushVersion: Rush_1.Rush.version
        };
        const logData2 = {
            name: 'testData2',
            duration: 100,
            result: 'Failed',
            timestamp: new Date().getTime(),
            platform: process.platform,
            rushVersion: Rush_1.Rush.version
        };
        telemetry.log(logData1);
        telemetry.log(logData2);
        expect(telemetry.store).toEqual([logData1, logData2]);
    });
    it('does not add data to store if telemetry is not enabled', () => {
        const filename = path.resolve(path.join(__dirname, './telemetry/telemetryNotEnabled.json'));
        const rushConfig = RushConfiguration_1.RushConfiguration.loadFromConfigurationFile(filename);
        const telemetry = new Telemetry_1.Telemetry(rushConfig);
        const logData = {
            name: 'testData',
            duration: 100,
            result: 'Succeeded',
            timestamp: new Date().getTime(),
            platform: process.platform,
            rushVersion: Rush_1.Rush.version
        };
        telemetry.log(logData);
        expect(telemetry.store).toEqual([]);
    });
    it('deletes data after flush', () => {
        const filename = path.resolve(path.join(__dirname, './telemetry/telemetryEnabled.json'));
        const rushConfig = RushConfiguration_1.RushConfiguration.loadFromConfigurationFile(filename);
        const telemetry = new Telemetry_1.Telemetry(rushConfig);
        const logData = {
            name: 'testData1',
            duration: 100,
            result: 'Succeeded',
            timestamp: new Date().getTime(),
            platform: process.platform,
            rushVersion: Rush_1.Rush.version
        };
        telemetry.log(logData);
        let logFile;
        let dataToWrite;
        telemetry.flush((file, data) => {
            logFile = file;
            dataToWrite = data;
        });
        expect(logFile.match(/telemetry_.*\.json/)).toBeDefined();
        expect(dataToWrite).toEqual(JSON.stringify([logData]));
        expect(telemetry.store).toEqual([]);
    });
    it('populates default fields', () => {
        const filename = path.resolve(path.join(__dirname, './telemetry/telemetryEnabled.json'));
        const rushConfig = RushConfiguration_1.RushConfiguration.loadFromConfigurationFile(filename);
        const telemetry = new Telemetry_1.Telemetry(rushConfig);
        const logData = {
            name: 'testData1',
            duration: 100,
            result: 'Succeeded'
        };
        telemetry.log(logData);
        const result = telemetry.store[0];
        expect(result.platform).toEqual(process.platform);
        expect(result.rushVersion).toEqual(Rush_1.Rush.version);
        expect(result.timestamp).toBeDefined();
    });
});
