// @microsoft/rush-lib/lib/cli/logic/InstallManager.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const glob = require("glob");
const colors = require("colors");
const fetch = require("node-fetch");
const HttpsProxyAgent = require("https-proxy-agent");
const os = require("os");
const path = require("path");
const fsx = require("fs-extra");
const semver = require("semver");
const tar = require("tar");
const wordwrap = require("wordwrap");
const globEscape = require("glob-escape");
const node_core_library_1 = require("@microsoft/node-core-library");
const ApprovedPackagesChecker_1 = require("../logic/ApprovedPackagesChecker");
const GitPolicy_1 = require("../logic/GitPolicy");
const LastInstallFlag_1 = require("../../utilities/LastInstallFlag");
const LinkManagerFactory_1 = require("../logic/LinkManagerFactory");
const RushConstants_1 = require("../../RushConstants");
const ShrinkwrapFileFactory_1 = require("../logic/ShrinkwrapFileFactory");
const Stopwatch_1 = require("../../utilities/Stopwatch");
const Utilities_1 = require("../../utilities/Utilities");
const Rush_1 = require("../../Rush");
const AlreadyReportedError_1 = require("../../utilities/AlreadyReportedError");
const MAX_INSTALL_ATTEMPTS = 5;
const wrap = wordwrap.soft(Utilities_1.Utilities.getConsoleWidth());
/**
 * This class implements common logic between "rush install" and "rush update".
 */
class InstallManager {
    /**
     * Returns a map of all direct dependencies that only have a single semantic version specifier
     */
    static collectImplicitlyPreferredVersions(rushConfiguration) {
        // First, collect all the direct dependencies of all local projects, and their versions:
        // direct dependency name --> set of version specifiers
        const versionsForDependencies = new Map();
        rushConfiguration.projects.forEach((project) => {
            InstallManager._collectVersionsForDependencies(versionsForDependencies, project.packageJson.dependencies, project.cyclicDependencyProjects, rushConfiguration);
            InstallManager._collectVersionsForDependencies(versionsForDependencies, project.packageJson.devDependencies, project.cyclicDependencyProjects, rushConfiguration);
        });
        // If any dependency has more than one version, then filter it out (since we don't know which version
        // should be preferred).  What remains will be the list of preferred dependencies.
        // dependency --> version specifier
        const implicitlyPreferred = new Map();
        versionsForDependencies.forEach((versions, dep) => {
            if (versions.size === 1) {
                const version = versions.values().next().value;
                implicitlyPreferred.set(dep, version);
            }
        });
        return implicitlyPreferred;
    }
    // Helper for collectImplicitlyPreferredVersions()
    static _updateVersionsForDependencies(versionsForDependencies, dependency, version) {
        if (!versionsForDependencies.has(dependency)) {
            versionsForDependencies.set(dependency, new Set());
        }
        versionsForDependencies.get(dependency).add(version);
    }
    // Helper for collectImplicitlyPreferredVersions()
    static _collectVersionsForDependencies(versionsForDependencies, dependencies, cyclicDependencies, rushConfiguration) {
        const allowedAlternativeVersions = rushConfiguration.commonVersions.allowedAlternativeVersions;
        if (dependencies) {
            Object.keys(dependencies).forEach((dependency) => {
                const versionSpecifier = dependencies[dependency];
                const alternativesForThisDependency = allowedAlternativeVersions.get(dependency) || [];
                // For each dependency, collectImplicitlyPreferredVersions() is collecting the set of all version specifiers
                // that appear across the repo.  If there is only one version specifier, then that's the "preferred" one.
                // However, there are a few cases where additional version specifiers can be safely ignored.
                let ignoreVersion = false;
                // 1. If the version specifier was listed in "allowedAlternativeVersions", then it's never a candidate.
                //    (Even if it's the only version specifier anywhere in the repo, we still ignore it, because
                //    otherwise the rule would be difficult to explain.)
                if (alternativesForThisDependency.indexOf(versionSpecifier) > 0) {
                    ignoreVersion = true;
                }
                else {
                    // Is it a local project?
                    const localProject = rushConfiguration.getProjectByName(dependency);
                    if (localProject) {
                        // 2. If it's a symlinked local project, then it's not a candidate, because the package manager will
                        //    never even see it.
                        // However there are two ways that a local project can NOT be symlinked:
                        // - if the local project doesn't satisfy the referenced semver specifier; OR
                        // - if the local project was specified in "cyclicDependencyProjects" in rush.json
                        if (semver.satisfies(localProject.packageJson.version, versionSpecifier)
                            && !cyclicDependencies.has(dependency)) {
                            ignoreVersion = true;
                        }
                    }
                }
                if (!ignoreVersion) {
                    InstallManager._updateVersionsForDependencies(versionsForDependencies, dependency, versionSpecifier);
                }
            });
        }
    }
    get commonNodeModulesMarker() {
        return this._commonNodeModulesMarker;
    }
    constructor(rushConfiguration, purgeManager) {
        this._rushConfiguration = rushConfiguration;
        this._commonTempFolderRecycler = purgeManager.commonTempFolderRecycler;
        this._commonNodeModulesMarker = new LastInstallFlag_1.LastInstallFlag(this._rushConfiguration.commonTempFolder, {
            node: process.versions.node,
            packageManager: rushConfiguration.packageManager,
            packageManagerVersion: rushConfiguration.packageManagerToolVersion
        });
    }
    doInstall(options) {
        return Promise.resolve().then(() => {
            // Check the policies
            if (!options.bypassPolicy) {
                if (!GitPolicy_1.GitPolicy.check(this._rushConfiguration)) {
                    throw new AlreadyReportedError_1.AlreadyReportedError();
                }
                ApprovedPackagesChecker_1.ApprovedPackagesChecker.rewriteConfigFiles(this._rushConfiguration);
            }
            // Ensure that the package manager is installed
            return this.ensureLocalPackageManager()
                .then(() => {
                let shrinkwrapFile = undefined;
                // (If it's a full update, then we ignore the shrinkwrap from Git since it will be overwritten)
                if (!options.fullUpgrade) {
                    try {
                        shrinkwrapFile = ShrinkwrapFileFactory_1.ShrinkwrapFileFactory.getShrinkwrapFile(this._rushConfiguration.packageManager, this._rushConfiguration.committedShrinkwrapFilename);
                    }
                    catch (ex) {
                        console.log();
                        console.log('Unable to load the shrinkwrap file: ' + ex.message);
                        if (!options.allowShrinkwrapUpdates) {
                            console.log();
                            console.log(colors.red('You need to run "rush update" to fix this problem'));
                            throw new AlreadyReportedError_1.AlreadyReportedError();
                        }
                        shrinkwrapFile = undefined;
                    }
                }
                const shrinkwrapIsUpToDate = this._createTempModulesAndCheckShrinkwrap(shrinkwrapFile)
                    && !options.recheckShrinkwrap;
                if (!shrinkwrapIsUpToDate) {
                    if (!options.allowShrinkwrapUpdates) {
                        console.log();
                        console.log(colors.red('The shrinkwrap file is out of date.  You need to run "rush update".'));
                        throw new AlreadyReportedError_1.AlreadyReportedError();
                    }
                }
                return this._installCommonModules(shrinkwrapIsUpToDate, options.allowShrinkwrapUpdates)
                    .then(() => {
                    if (!options.noLink) {
                        const linkManager = LinkManagerFactory_1.LinkManagerFactory.getLinkManager(this._rushConfiguration);
                        return linkManager.createSymlinksForProjects(false);
                    }
                    else {
                        console.log(os.EOL
                            + colors.yellow('Since "--no-link" was specified, you will need to run "rush link" manually.'));
                    }
                });
            });
        });
    }
    /**
     * If the "(p)npm-local" symlink hasn't been set up yet, this creates it, installing the
     * specified (P)npm version in the user's home directory if needed.
     */
    ensureLocalPackageManager() {
        // Example: "C:\Users\YourName\.rush"
        const rushUserFolder = this._rushConfiguration.rushUserFolder;
        if (!fsx.existsSync(rushUserFolder)) {
            console.log('Creating ' + rushUserFolder);
            fsx.mkdirSync(rushUserFolder);
        }
        const packageManager = this._rushConfiguration.packageManager;
        const packageManagerVersion = this._rushConfiguration.packageManagerToolVersion;
        const packageManagerAndVersion = `${packageManager}-${packageManagerVersion}`;
        // Example: "C:\Users\YourName\.rush\pnpm-1.2.3"
        const packageManagerToolFolder = path.join(rushUserFolder, packageManagerAndVersion);
        const packageManagerMarker = new LastInstallFlag_1.LastInstallFlag(packageManagerToolFolder, {
            node: process.versions.node
        });
        console.log(`Trying to acquire lock for ${packageManagerAndVersion}`);
        return node_core_library_1.LockFile.acquire(rushUserFolder, packageManagerAndVersion).then((lock) => {
            console.log(`Acquired lock for ${packageManagerAndVersion}`);
            if (!packageManagerMarker.isValid() || lock.dirtyWhenAcquired) {
                console.log(colors.bold(`Installing ${packageManager} version ${packageManagerVersion}${os.EOL}`));
                // note that this will remove the last-install flag from the directory
                Utilities_1.Utilities.installPackageInDirectory(packageManagerToolFolder, packageManager, this._rushConfiguration.packageManagerToolVersion, `${packageManager}-local-install`, MAX_INSTALL_ATTEMPTS);
                console.log(`Successfully installed ${packageManager} version ${packageManagerVersion}`);
            }
            else {
                console.log(`Found ${packageManager} version ${packageManagerVersion} in ${packageManagerToolFolder}`);
            }
            packageManagerMarker.create();
            // Example: "C:\MyRepo\common\temp"
            if (!fsx.existsSync(this._rushConfiguration.commonTempFolder)) {
                fsx.mkdirsSync(this._rushConfiguration.commonTempFolder);
            }
            // Example: "C:\MyRepo\common\temp\pnpm-local"
            const localPackageManagerToolFolder = path.join(this._rushConfiguration.commonTempFolder, `${packageManager}-local`);
            console.log(os.EOL + 'Symlinking "' + localPackageManagerToolFolder + '"');
            console.log('  --> "' + packageManagerToolFolder + '"');
            // We cannot use fsx.existsSync() to test the existence of a symlink, because it will
            // return false for broken symlinks.  There is no way to test without catching an exception.
            try {
                fsx.unlinkSync(localPackageManagerToolFolder);
            }
            catch (error) {
                if (error.code !== 'ENOENT') {
                    throw error;
                }
            }
            fsx.symlinkSync(packageManagerToolFolder, localPackageManagerToolFolder, 'junction');
            lock.release();
        });
    }
    /**
     * Regenerates the common/package.json and all temp_modules projects.
     * If shrinkwrapFile is provided, this function also validates whether it contains
     * everything we need to install and returns true if so; in all other cases,
     * the return value is false.
     */
    _createTempModulesAndCheckShrinkwrap(shrinkwrapFile) {
        const stopwatch = Stopwatch_1.Stopwatch.start();
        // Example: "C:\MyRepo\common\temp\projects"
        const tempProjectsFolder = path.join(this._rushConfiguration.commonTempFolder, RushConstants_1.RushConstants.rushTempProjectsFolderName);
        console.log(os.EOL + colors.bold('Updating temp projects in ' + tempProjectsFolder));
        Utilities_1.Utilities.createFolderWithRetry(tempProjectsFolder);
        // We will start with the assumption that it's valid, and then set it to false if
        // any of the checks fail
        let shrinkwrapIsUpToDate = true;
        if (!shrinkwrapFile) {
            shrinkwrapIsUpToDate = false;
        }
        // Find the implicitly preferred versions
        // These are any first-level dependencies for which we only consume a single version range
        // (e.g. every package that depends on react uses an identical specifier)
        const allPreferredVersions = InstallManager.collectImplicitlyPreferredVersions(this._rushConfiguration);
        // Add in the explicitly preferred versions.
        // Note that these take precedence over implicitly preferred versions.
        node_core_library_1.MapExtensions.mergeFromMap(allPreferredVersions, this._rushConfiguration.commonVersions.getAllPreferredVersions());
        if (shrinkwrapFile) {
            // Check any preferred dependencies first
            allPreferredVersions.forEach((version, dependency) => {
                if (!shrinkwrapFile.hasCompatibleTopLevelDependency(dependency, version)) {
                    console.log(colors.yellow(wrap(`${os.EOL}The NPM shrinkwrap file does not provide "${dependency}"`
                        + ` (${version}) required by the preferred versions from ` + RushConstants_1.RushConstants.commonVersionsFilename)));
                    shrinkwrapIsUpToDate = false;
                }
            });
            if (this._findOrphanedTempProjects(shrinkwrapFile)) {
                // If there are any orphaned projects, then "npm install" would fail because the shrinkwrap
                // contains references such as "resolved": "file:projects\\project1" that refer to nonexistent
                // file paths.
                shrinkwrapIsUpToDate = false;
            }
        }
        // Also copy down the committed .npmrc file, if there is one
        // "common\config\rush\.npmrc" --> "common\temp\.npmrc"
        const committedNpmrcPath = path.join(this._rushConfiguration.commonRushConfigFolder, '.npmrc');
        const tempNpmrcPath = path.join(this._rushConfiguration.commonTempFolder, '.npmrc');
        // ensure that we remove any old one that may be hanging around
        this._syncFile(committedNpmrcPath, tempNpmrcPath);
        // also, copy the pnpmfile.js if it exists
        if (this._rushConfiguration.packageManager === 'pnpm') {
            const committedPnpmFilePath = path.join(this._rushConfiguration.commonRushConfigFolder, RushConstants_1.RushConstants.pnpmFileFilename);
            const tempPnpmFilePath = path.join(this._rushConfiguration.commonTempFolder, RushConstants_1.RushConstants.pnpmFileFilename);
            // ensure that we remove any old one that may be hanging around
            this._syncFile(committedPnpmFilePath, tempPnpmFilePath);
        }
        const commonPackageJson = {
            dependencies: {},
            description: 'Temporary file generated by the Rush tool',
            name: 'rush-common',
            private: true,
            version: '0.0.0'
        };
        // Add any preferred versions to the top of the commonPackageJson
        // do this in alphabetical order for simpler debugging
        for (const dependency of Array.from(allPreferredVersions.keys()).sort()) {
            commonPackageJson.dependencies[dependency] = allPreferredVersions.get(dependency);
        }
        // To make the common/package.json file more readable, sort alphabetically
        // according to rushProject.tempProjectName instead of packageName.
        const sortedRushProjects = this._rushConfiguration.projects.slice(0);
        sortedRushProjects.sort((a, b) => a.tempProjectName.localeCompare(b.tempProjectName));
        for (const rushProject of sortedRushProjects) {
            const packageJson = rushProject.packageJson;
            // Example: "C:\MyRepo\common\temp\projects\my-project-2.tgz"
            const tarballFile = this._getTarballFilePath(rushProject);
            // Example: "my-project-2"
            const unscopedTempProjectName = rushProject.unscopedTempProjectName;
            // Example: dependencies["@rush-temp/my-project-2"] = "file:./projects/my-project-2.tgz"
            commonPackageJson.dependencies[rushProject.tempProjectName]
                = `file:./${RushConstants_1.RushConstants.rushTempProjectsFolderName}/${rushProject.unscopedTempProjectName}.tgz`;
            const tempPackageJson = {
                name: rushProject.tempProjectName,
                version: '0.0.0',
                private: true,
                dependencies: {}
            };
            // If there are any optional dependencies, copy them over directly
            if (packageJson.optionalDependencies) {
                tempPackageJson.optionalDependencies = packageJson.optionalDependencies;
            }
            // Collect pairs of (packageName, packageVersion) to be added as temp package dependencies
            const pairs = [];
            // If there are devDependencies, we need to merge them with the regular
            // dependencies.  If the same library appears in both places, then the
            // regular dependency takes precedence over the devDependency.
            // It also takes precedence over a duplicate in optionalDependencies,
            // but NPM will take care of that for us.  (Frankly any kind of duplicate
            // should be an error, but NPM is pretty lax about this.)
            if (packageJson.devDependencies) {
                for (const packageName of Object.keys(packageJson.devDependencies)) {
                    pairs.push({ packageName: packageName, packageVersion: packageJson.devDependencies[packageName] });
                }
            }
            if (packageJson.dependencies) {
                for (const packageName of Object.keys(packageJson.dependencies)) {
                    pairs.push({ packageName: packageName, packageVersion: packageJson.dependencies[packageName] });
                }
            }
            for (const pair of pairs) {
                // Is there a locally built Rush project that could satisfy this dependency?
                // If so, then we will symlink to the project folder rather than to common/temp/node_modules.
                // In this case, we don't want "npm install" to process this package, but we do need
                // to record this decision for "rush link" later, so we add it to a special 'rushDependencies' field.
                const localProject = this._rushConfiguration.getProjectByName(pair.packageName);
                if (localProject) {
                    // Don't locally link if it's listed in the cyclicDependencyProjects
                    if (!rushProject.cyclicDependencyProjects.has(pair.packageName)) {
                        // Also, don't locally link if the SemVer doesn't match
                        const localProjectVersion = localProject.packageJson.version;
                        if (semver.satisfies(localProjectVersion, pair.packageVersion)) {
                            // We will locally link this package
                            if (!tempPackageJson.rushDependencies) {
                                tempPackageJson.rushDependencies = {};
                            }
                            tempPackageJson.rushDependencies[pair.packageName] = pair.packageVersion;
                            continue;
                        }
                    }
                }
                // We will NOT locally link this package; add it as a regular dependency.
                tempPackageJson.dependencies[pair.packageName] = pair.packageVersion;
                if (shrinkwrapFile) {
                    if (!shrinkwrapFile.tryEnsureCompatibleDependency(pair.packageName, pair.packageVersion, rushProject.tempProjectName)) {
                        console.log(colors.yellow(wrap(`${os.EOL}The NPM shrinkwrap file is missing "${pair.packageName}"`
                            + ` (${pair.packageVersion}) required by "${rushProject.packageName}".`)));
                        shrinkwrapIsUpToDate = false;
                    }
                }
            }
            // NPM expects the root of the tarball to have a directory called 'package'
            const npmPackageFolder = 'package';
            // Example: "C:\MyRepo\common\temp\projects\my-project-2"
            const tempProjectFolder = path.join(this._rushConfiguration.commonTempFolder, RushConstants_1.RushConstants.rushTempProjectsFolderName, unscopedTempProjectName);
            // Example: "C:\MyRepo\common\temp\projects\my-project-2\package.json"
            const tempPackageJsonFilename = path.join(tempProjectFolder, RushConstants_1.RushConstants.packageJsonFilename);
            // we only want to overwrite the package if the existing tarball's package.json is different from tempPackageJson
            let shouldOverwrite = true;
            try {
                // if the tarball and the temp file still exist, then compare the contents
                if (fsx.existsSync(tarballFile) && fsx.existsSync(tempPackageJsonFilename)) {
                    // compare the extracted package.json with the one we are about to write
                    const oldBuffer = fsx.readFileSync(tempPackageJsonFilename);
                    const newBuffer = new Buffer(node_core_library_1.JsonFile.stringify(tempPackageJson));
                    if (Buffer.compare(oldBuffer, newBuffer) === 0) {
                        shouldOverwrite = false;
                    }
                }
            }
            catch (error) {
                // ignore the error, we will go ahead and create a new tarball
            }
            if (shouldOverwrite) {
                try {
                    // ensure the folder we are about to zip exists
                    Utilities_1.Utilities.createFolderWithRetry(tempProjectFolder);
                    // remove the old tarball & old temp package json, this is for any cases where new tarball creation
                    // fails, and the shouldOverwrite logic is messed up because the my-project-2\package.json
                    // exists and is updated, but the tarball is not accurate
                    fsx.removeSync(tarballFile);
                    fsx.removeSync(tempPackageJsonFilename);
                    // write the expected package.json file into the zip staging folder
                    node_core_library_1.JsonFile.save(tempPackageJson, tempPackageJsonFilename);
                    // create the new tarball
                    tar.create({
                        gzip: true,
                        file: tarballFile,
                        cwd: tempProjectFolder,
                        portable: true,
                        noMtime: true,
                        noPax: true,
                        sync: true,
                        prefix: npmPackageFolder
                    }, ['package.json']);
                    console.log(`Updating ${tarballFile}`);
                }
                catch (error) {
                    // delete everything in case of any error
                    fsx.removeSync(tarballFile);
                    fsx.removeSync(tempPackageJsonFilename);
                }
            }
        }
        // Example: "C:\MyRepo\common\temp\package.json"
        const commonPackageJsonFilename = path.join(this._rushConfiguration.commonTempFolder, RushConstants_1.RushConstants.packageJsonFilename);
        if (shrinkwrapFile) {
            // If we have a (possibly incomplete) shrinkwrap file, save it as the temporary file.
            shrinkwrapFile.save(this._rushConfiguration.tempShrinkwrapFilename);
            shrinkwrapFile.save(this._rushConfiguration.tempShrinkwrapPreinstallFilename);
        }
        else {
            // Otherwise delete the temporary file
            fsx.removeSync(this._rushConfiguration.tempShrinkwrapFilename);
        }
        // Don't update the file timestamp unless the content has changed, since "rush install"
        // will consider this timestamp
        node_core_library_1.JsonFile.save(commonPackageJson, commonPackageJsonFilename, { onlyIfChanged: true });
        stopwatch.stop();
        console.log(`Finished creating temporary modules (${stopwatch.toString()})`);
        return shrinkwrapIsUpToDate;
    }
    /**
     * Runs "npm install" in the common folder.
     */
    _installCommonModules(shrinkwrapIsUpToDate, allowShrinkwrapUpdates) {
        return Promise.resolve().then(() => {
            console.log(os.EOL + colors.bold('Checking node_modules in ' + this._rushConfiguration.commonTempFolder)
                + os.EOL);
            const commonNodeModulesFolder = path.join(this._rushConfiguration.commonTempFolder, 'node_modules');
            // This marker file indicates that the last "rush install" completed successfully
            const markerFileExistedAndWasValidAtStart = this._commonNodeModulesMarker.isValid();
            // If "--clean" or "--full-clean" was specified, or if the last install was interrupted,
            // then we will need to delete the node_modules folder.  Otherwise, we can do an incremental
            // install.
            const deleteNodeModules = !markerFileExistedAndWasValidAtStart;
            // Based on timestamps, can we skip this install entirely?
            if (shrinkwrapIsUpToDate && !deleteNodeModules) {
                const potentiallyChangedFiles = [];
                // Consider the timestamp on the node_modules folder; if someone tampered with it
                // or deleted it entirely, then we can't skip this install
                potentiallyChangedFiles.push(commonNodeModulesFolder);
                // Additionally, if they pulled an updated npm-shrinkwrap.json file from Git,
                // then we can't skip this install
                potentiallyChangedFiles.push(this._rushConfiguration.committedShrinkwrapFilename);
                if (this._rushConfiguration.packageManager === 'pnpm') {
                    // If the repo is using pnpmfile.js, consider that also
                    const pnpmFileFilename = path.join(this._rushConfiguration.commonRushConfigFolder, RushConstants_1.RushConstants.pnpmFileFilename);
                    if (fsx.existsSync(pnpmFileFilename)) {
                        potentiallyChangedFiles.push(pnpmFileFilename);
                    }
                }
                // Also consider timestamps for all the temp tarballs. (createTempModulesAndCheckShrinkwrap() will
                // carefully preserve these timestamps unless something has changed.)
                // Example: "C:\MyRepo\common\temp\projects\my-project-2.tgz"
                potentiallyChangedFiles.push(...this._rushConfiguration.projects.map(x => {
                    return this._getTarballFilePath(x);
                }));
                // NOTE: If commonNodeModulesMarkerFilename (or any of the potentiallyChangedFiles) does not
                // exist, then isFileTimestampCurrent() returns false.
                if (Utilities_1.Utilities.isFileTimestampCurrent(this._commonNodeModulesMarker.path, potentiallyChangedFiles)) {
                    // Nothing to do, because everything is up to date according to time stamps
                    return;
                }
            }
            return this._checkIfReleaseIsPublished()
                .catch((error) => {
                // If the user is working in an environment that can't reach the registry,
                // don't bother them with errors.
                return undefined;
            }).then((publishedRelease) => {
                if (publishedRelease === false) {
                    console.log(colors.yellow('Warning: This release of the Rush tool was unpublished; it may be unstable.'));
                }
                // Since we're going to be tampering with common/node_modules, delete the "rush link" flag file if it exists;
                // this ensures that a full "rush link" is required next time
                Utilities_1.Utilities.deleteFile(this._rushConfiguration.rushLinkJsonFilename);
                // Delete the successful install file to indicate the install transaction has started
                this._commonNodeModulesMarker.clear();
                // NOTE: The PNPM store is supposed to be transactionally safe, so we don't delete it automatically.
                // The user must request that via the command line.
                if (deleteNodeModules) {
                    if (this._rushConfiguration.packageManager === 'npm') {
                        console.log(`Deleting the "npm-cache" folder`);
                        // This is faster and more thorough than "npm cache clean"
                        this._commonTempFolderRecycler.moveFolder(this._rushConfiguration.npmCacheFolder);
                        console.log(`Deleting the "npm-tmp" folder`);
                        this._commonTempFolderRecycler.moveFolder(this._rushConfiguration.npmTmpFolder);
                    }
                }
                // Example: "C:\MyRepo\common\temp\npm-local\node_modules\.bin\npm"
                const packageManagerFilename = this._rushConfiguration.packageManagerToolFilename;
                // Is there an existing "node_modules" folder to consider?
                if (fsx.existsSync(commonNodeModulesFolder)) {
                    // Should we delete the entire "node_modules" folder?
                    if (deleteNodeModules) {
                        // YES: Delete "node_modules"
                        // Explain to the user why we are hosing their node_modules folder
                        console.log('Deleting files from ' + commonNodeModulesFolder);
                        this._commonTempFolderRecycler.moveFolder(commonNodeModulesFolder);
                        Utilities_1.Utilities.createFolderWithRetry(commonNodeModulesFolder);
                    }
                    else {
                        // NO: Prepare to do an incremental install in the "node_modules" folder
                        // note: it is not necessary to run "prune" with pnpm
                        if (this._rushConfiguration.packageManager === 'npm') {
                            console.log(`Running "${this._rushConfiguration.packageManager} prune"`
                                + ` in ${this._rushConfiguration.commonTempFolder}`);
                            const args = ['prune'];
                            this._pushConfigurationArgs(args);
                            Utilities_1.Utilities.executeCommandWithRetry(MAX_INSTALL_ATTEMPTS, packageManagerFilename, args, this._rushConfiguration.commonTempFolder);
                            // Delete the (installed image of) the temp projects, since "npm install" does not
                            // detect changes for "file:./" references.
                            // We recognize the temp projects by their names, which always start with "rush-".
                            // Example: "C:\MyRepo\common\temp\node_modules\@rush-temp"
                            const pathToDeleteWithoutStar = path.join(commonNodeModulesFolder, RushConstants_1.RushConstants.rushTempNpmScope);
                            console.log(`Deleting ${pathToDeleteWithoutStar}\\*`);
                            // Glob can't handle Windows paths
                            const normalizedpathToDeleteWithoutStar = node_core_library_1.Text.replaceAll(pathToDeleteWithoutStar, '\\', '/');
                            // Example: "C:/MyRepo/common/temp/node_modules/@rush-temp/*"
                            for (const tempModulePath of glob.sync(globEscape(normalizedpathToDeleteWithoutStar) + '/*')) {
                                // We could potentially use AsyncRecycler here, but in practice these folders tend
                                // to be very small
                                Utilities_1.Utilities.dangerouslyDeletePath(tempModulePath);
                            }
                        }
                    }
                }
                // Run "npm install" in the common folder
                // NOTE:
                // we do NOT install optional dependencies for Rush, as it seems that optional dependencies do not
                // work properly with shrinkwrap. Consider the "fsevents" package. This is a Mac specific package
                // which is an optional second-order dependency. Optional dependencies work by attempting to install
                // the package, but removes the package if the install failed.
                // This means that someone running generate on a Mac WILL have fsevents included in their shrinkwrap.
                // When someone using Windows attempts to install from the shrinkwrap, the install will fail.
                //
                // If someone generates the shrinkwrap using Windows, then fsevents will NOT be listed in the shrinkwrap.
                // When someone using Mac attempts to install from the shrinkwrap, (as of NPM 4), they will NOT have the
                // optional dependency installed.
                //
                // One possible solution would be to have the shrinkwrap include information about whether the dependency
                // is optional or not, but it does not appear to do so. Also, this would result in strange behavior where
                // people would have different node_modules based on their system.
                const installArgs = ['install', '--no-optional'];
                this._pushConfigurationArgs(installArgs);
                console.log(os.EOL + colors.bold(`Running "${this._rushConfiguration.packageManager} install" in`
                    + ` ${this._rushConfiguration.commonTempFolder}`) + os.EOL);
                Utilities_1.Utilities.executeCommandWithRetry(MAX_INSTALL_ATTEMPTS, packageManagerFilename, installArgs, this._rushConfiguration.commonTempFolder, undefined, false, () => {
                    if (this._rushConfiguration.packageManager === 'pnpm') {
                        // If there is a failure in pnpm, it is possible that it left the
                        // store in a bad state. Therefore, we should clean out the store
                        // before attempting the install again.
                        console.log(colors.yellow(`Deleting the "node_modules" folder`));
                        this._commonTempFolderRecycler.moveFolder(commonNodeModulesFolder);
                        console.log(colors.yellow(`Deleting the "pnpm-store" folder`));
                        this._commonTempFolderRecycler.moveFolder(this._rushConfiguration.pnpmStoreFolder);
                        Utilities_1.Utilities.createFolderWithRetry(commonNodeModulesFolder);
                    }
                });
                if (this._rushConfiguration.packageManager === 'npm') {
                    console.log(os.EOL + colors.bold('Running "npm shrinkwrap"...'));
                    const npmArgs = ['shrinkwrap'];
                    this._pushConfigurationArgs(npmArgs);
                    Utilities_1.Utilities.executeCommand(this._rushConfiguration.packageManagerToolFilename, npmArgs, this._rushConfiguration.commonTempFolder);
                    console.log('"npm shrinkwrap" completed' + os.EOL);
                    this._fixupNpm5Regression();
                }
                if (allowShrinkwrapUpdates && !shrinkwrapIsUpToDate) {
                    // Copy (or delete) common\temp\shrinkwrap.yaml --> common\config\rush\shrinkwrap.yaml
                    this._syncFile(this._rushConfiguration.tempShrinkwrapFilename, this._rushConfiguration.committedShrinkwrapFilename);
                }
                else {
                    // TODO: Validate whether the package manager updated it in a nontrivial way
                }
                // Finally, create the marker file to indicate a successful install
                this._commonNodeModulesMarker.create();
                console.log('');
            });
        });
    }
    _checkIfReleaseIsPublished() {
        return Promise.resolve().then(() => {
            const lastCheckFile = path.join(this._rushConfiguration.rushUserFolder, 'rush-' + Rush_1.Rush.version, 'last-check.flag');
            if (fsx.existsSync(lastCheckFile)) {
                let cachedResult = undefined;
                try {
                    // NOTE: mtimeMs is not supported yet in NodeJS 6.x
                    const nowMs = new Date().getTime();
                    const ageMs = nowMs - fsx.statSync(lastCheckFile).mtime.getTime();
                    const HOUR = 60 * 60 * 1000;
                    // Is the cache too old?
                    if (ageMs < 24 * HOUR) {
                        // No, read the cached result
                        cachedResult = node_core_library_1.JsonFile.load(lastCheckFile);
                    }
                }
                catch (e) {
                    // Unable to parse file
                }
                if (cachedResult === 'error') {
                    return Promise.reject(new Error('Unable to contact server'));
                }
                if (cachedResult === true || cachedResult === false) {
                    return cachedResult;
                }
            }
            // Before we start the network operation, record a failed state.  If the process exits for some reason,
            // this will record the error.  It will also update the timestamp to prevent other Rush instances
            // from attempting to update the file.
            fsx.mkdirsSync(path.dirname(lastCheckFile));
            node_core_library_1.JsonFile.save('error', lastCheckFile);
            // For this check we use the official registry, not the private registry
            return this._queryIfReleaseIsPublished('https://registry.npmjs.org:443')
                .then((publishedRelease) => {
                // Cache the result
                fsx.mkdirsSync(path.dirname(lastCheckFile));
                node_core_library_1.JsonFile.save(publishedRelease, lastCheckFile);
                return publishedRelease;
            })
                .catch((error) => {
                fsx.mkdirsSync(path.dirname(lastCheckFile));
                node_core_library_1.JsonFile.save('error', lastCheckFile);
                return Promise.reject(error);
            });
        });
    }
    _queryIfReleaseIsPublished(registryUrl) {
        let queryUrl = registryUrl;
        if (queryUrl[-1] !== '/') {
            queryUrl += '/';
        }
        // Note that the "@" symbol does not normally get URL-encoded
        queryUrl += RushConstants_1.RushConstants.rushPackageName.replace('/', '%2F');
        const userAgent = `pnpm/? npm/? node/${process.version} ${os.platform()} ${os.arch()}`;
        const headers = new fetch.Headers();
        headers.append('user-agent', userAgent);
        headers.append('accept', 'application/vnd.npm.install-v1+json; q=1.0, application/json; q=0.8, */*');
        let agent = undefined;
        if (process.env.HTTP_PROXY) {
            agent = new HttpsProxyAgent(process.env.HTTP_PROXY);
        }
        return fetch.default(queryUrl, {
            headers: headers,
            agent: agent
        })
            .then((response) => {
            if (!response.ok) {
                return Promise.reject(new Error('Failed to query'));
            }
            return response
                .json()
                .then((data) => {
                let url;
                try {
                    if (!data.versions[Rush_1.Rush.version]) {
                        // Version was not published
                        return false;
                    }
                    url = data.versions[Rush_1.Rush.version].dist.tarball;
                    if (!url) {
                        return Promise.reject(new Error(`URL not found`));
                    }
                }
                catch (e) {
                    return Promise.reject(new Error('Error parsing response'));
                }
                // Make sure the tarball wasn't deleted from the CDN
                headers.set('accept', '*/*');
                return fetch.default(url, {
                    headers: headers,
                    agent: agent
                })
                    .then((response2) => {
                    if (!response2.ok) {
                        if (response2.status === 404) {
                            return false;
                        }
                        else {
                            return Promise.reject(new Error('Failed to fetch'));
                        }
                    }
                    return true;
                });
            });
        });
    }
    /**
     * Used when invoking the NPM tool.  Appends the common configuration options
     * to the command-line.
     */
    _pushConfigurationArgs(args) {
        if (this._rushConfiguration.packageManager === 'npm') {
            args.push('--cache', this._rushConfiguration.npmCacheFolder);
            args.push('--tmp', this._rushConfiguration.npmTmpFolder);
        }
        else if (this._rushConfiguration.packageManager === 'pnpm') {
            args.push('--store', this._rushConfiguration.pnpmStoreFolder);
            // we are using the --no-lock flag for now, which unfortunately prints a warning, but should be OK
            // since rush already has its own install lock file which will invalidate the cache for us.
            // we theoretically could use the lock file, but we would need to clean the store if the
            // lockfile existed, otherwise PNPM would hang indefinitely. it is simpler to rely on Rush's
            // last install flag, which encapsulates the entire installation
            args.push('--no-lock');
        }
    }
    /**
     * Copies the file "sourcePath" to "targetPath", overwriting the target file location.
     * If the source file does not exist, then the target file is deleted.
     */
    _syncFile(sourcePath, targetPath) {
        if (fsx.existsSync(sourcePath)) {
            console.log('Updating ' + targetPath);
            fsx.copySync(sourcePath, targetPath);
        }
        else {
            if (fsx.existsSync(targetPath)) {
                console.log('Deleting ' + targetPath);
                fsx.unlinkSync(targetPath);
            }
        }
    }
    /**
     * Gets the path to the tarball
     * Example: "C:\MyRepo\common\temp\projects\my-project-2.tgz"
     */
    _getTarballFilePath(project) {
        return path.join(this._rushConfiguration.commonTempFolder, RushConstants_1.RushConstants.rushTempProjectsFolderName, `${project.unscopedTempProjectName}.tgz`);
    }
    /**
     * This is a workaround for a bug introduced in NPM 5 (and still unfixed as of NPM 5.5.1):
     * https://github.com/npm/npm/issues/19006
     *
     * The regression is that "npm install" sets the package.json "version" field for the
     * @rush-temp projects to a value like "file:projects/example.tgz", when it should be "0.0.0".
     * This causes "rush link" to fail later, when read-package-tree tries to parse the bad version.
     * The error looks like this:
     *
     * ERROR: Failed to parse package.json for foo: Invalid version: "file:projects/example.tgz"
     *
     * Our workaround is to rewrite the package.json files for each of the @rush-temp projects
     * in the node_modules folder, after "npm install" completes.
     */
    _fixupNpm5Regression() {
        const pathToDeleteWithoutStar = path.join(this._rushConfiguration.commonTempFolder, 'node_modules', RushConstants_1.RushConstants.rushTempNpmScope);
        // Glob can't handle Windows paths
        const normalizedpathToDeleteWithoutStar = node_core_library_1.Text.replaceAll(pathToDeleteWithoutStar, '\\', '/');
        let anyChanges = false;
        // Example: "C:/MyRepo/common/temp/node_modules/@rush-temp/*/package.json"
        for (const packageJsonPath of glob.sync(globEscape(normalizedpathToDeleteWithoutStar) + '/*/package.json')) {
            // Example: "C:/MyRepo/common/temp/node_modules/@rush-temp/example/package.json"
            const packageJsonObject = node_core_library_1.JsonFile.load(packageJsonPath);
            // The temp projects always use "0.0.0" as their version
            packageJsonObject.version = '0.0.0';
            if (node_core_library_1.JsonFile.save(packageJsonObject, packageJsonPath, { onlyIfChanged: true })) {
                anyChanges = true;
            }
        }
        if (anyChanges) {
            console.log(os.EOL + colors.yellow(wrap(`Applied workaround for NPM 5 bug`)) + os.EOL);
        }
    }
    /**
     * Checks for temp projects that exist in the shrinkwrap file, but don't exist
     * in rush.json.  This might occur, e.g. if a project was recently deleted or renamed.
     *
     * @returns true if orphans were found, or false if everything is okay
     */
    _findOrphanedTempProjects(shrinkwrapFile) {
        // We can recognize temp projects because they are under the "@rush-temp" NPM scope.
        for (const tempProjectName of shrinkwrapFile.getTempProjectNames()) {
            if (!this._rushConfiguration.findProjectByTempName(tempProjectName)) {
                console.log(os.EOL + colors.yellow(wrap(`Your NPM shrinkwrap file references a project "${tempProjectName}" which no longer exists.`))
                    + os.EOL);
                return true; // found one
            }
        }
        return false; // none found
    }
}
exports.InstallManager = InstallManager;
