// @microsoft/rush-lib/lib/cli/actions/BaseRushAction.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const colors = require("colors");
const os = require("os");
const path = require("path");
const ts_command_line_1 = require("@microsoft/ts-command-line");
const node_core_library_1 = require("@microsoft/node-core-library");
const EventHooksManager_1 = require("../logic/EventHooksManager");
/**
 * The base Rush action that all Rush actions should extend.
 */
class BaseRushAction extends ts_command_line_1.CommandLineAction {
    get rushConfiguration() {
        return this._parser.rushConfiguration;
    }
    get parser() {
        return this._parser;
    }
    constructor(options) {
        super(options);
        this._parser = options.parser;
        this._safeForSimultaneousRushProcesses = !!options.safeForSimultaneousRushProcesses;
    }
    onExecute() {
        // Defensively set the exit code to 1 so if Rush crashes for whatever reason, we'll have a nonzero exit code.
        // For example, NodeJS currently has the inexcusable design of terminating with zero exit code when
        // there is an uncaught promise exception.  This will supposedly be fixed in NodeJS 9.
        // Ideally we should do this for all the Rush actions, but "rush build" is the most critical one
        // -- if it falsely appears to succeed, we could merge bad PRs, publish empty packages, etc.
        process.exitCode = 1;
        this._ensureEnvironment();
        if (!this._safeForSimultaneousRushProcesses) {
            if (!node_core_library_1.LockFile.tryAcquire(this.rushConfiguration.commonTempFolder, 'rush')) {
                console.log(colors.red(`Another rush command is already running in this repository.`));
                process.exit(1);
            }
        }
        console.log(`Starting "rush ${this.actionName}"${os.EOL}`);
        return this.run().then(() => {
            // If we make it here, everything went fine, so reset the exit code back to 0
            process.exitCode = 0;
        });
    }
    get eventHooksManager() {
        if (!this._eventHooksManager) {
            this._eventHooksManager = new EventHooksManager_1.EventHooksManager(this.rushConfiguration.eventHooks, this.rushConfiguration.commonTempFolder);
        }
        return this._eventHooksManager;
    }
    _ensureEnvironment() {
        /* tslint:disable-next-line:no-string-literal */
        let environmentPath = process.env['PATH'];
        environmentPath = path.join(this.rushConfiguration.commonTempFolder, 'node_modules', '.bin') +
            path.delimiter + environmentPath;
        /* tslint:disable-next-line:no-string-literal */
        process.env['PATH'] = environmentPath;
    }
}
exports.BaseRushAction = BaseRushAction;
