// @microsoft/sp-webpart-base/lib/components/placeHolder/PlaceholderSpinner.js
import * as tslib_1 from "tslib";
import { sealed } from '@microsoft/decorators';
import * as React from 'react';
import { Overlay, Spinner, SpinnerType } from '@microsoft/office-ui-fabric-react-bundle';
import './Placeholder.scss';
var PlaceholderSpinner = (function (_super) {
    tslib_1.__extends(PlaceholderSpinner, _super);
    function PlaceholderSpinner(props) {
        var _this = _super.call(this, props) || this;
        _this.showSpinner = _this.showSpinner.bind(_this);
        _this.state = { showSpinner: !props.hideSpinner };
        return _this;
    }
    PlaceholderSpinner.prototype.componentDidMount = function () {
        this._isMounted = true;
    };
    PlaceholderSpinner.prototype.render = function () {
        if (this.state.removePlaceholder) {
            return null; 
        }
        return (React.createElement(Overlay, { className: 'Placeholder-overlay' },
            React.createElement("div", { className: 'Placeholder-spinnerContainer' }, this.state.showSpinner && React.createElement(Spinner, { type: SpinnerType.large, label: this.props.label }))));
    };
    PlaceholderSpinner.prototype.componentWillUnmount = function () {
        this._isMounted = false;
    };
    PlaceholderSpinner.prototype.showSpinner = function (showSpinner) {
        if (this._isMounted) {
            this.setState({ showSpinner: showSpinner });
        }
    };
    PlaceholderSpinner.prototype.clearPlaceholder = function () {
        if (this._isMounted) {
            this.setState({ removePlaceholder: true });
        }
    };
    PlaceholderSpinner = tslib_1.__decorate([
        sealed
    ], PlaceholderSpinner);
    return PlaceholderSpinner;
}(React.Component));
export default PlaceholderSpinner;
