// @microsoft/sp-application-base/lib/pageChrome/SPPageChrome.js
import { _SPKillSwitch, _SPFlight, Guid } from '@microsoft/sp-core-library';
import { _SPLoaderFlights } from '@microsoft/sp-loader';
import SPThemeProvider from './SPThemeProvider';
import SPGlobalErrorHandler from '../error/SPGlobalErrorHandler';
import DialogManagerConfiguration from '../extensibility/DialogManagerConfiguration';
import { _PerformanceLogger } from '@ms/sp-telemetry';
import styles from './SPPageChrome.module.scss';
var SPPageChrome = (function () {
    function SPPageChrome(serviceScope) {
        this._isDisposed = false;
        this._createPageChrome();
        DialogManagerConfiguration.instance.initialize(this.dialogContainerDiv);
        var guid = Guid.parse('48CB5A9D-4E94-40F0-B51F-24480C28E565');
        if (!_SPKillSwitch.isActivated(guid, '7/13/2016', 'EnableTheming')) {
            this._spThemeProvider = new SPThemeProvider(serviceScope);
            if (_SPKillSwitch.isActivated(SPPageChrome.DEFER_LOAD_THEME_KILL_SWITCH, '04/04/2018', 'DeferLoadTheme')) {
                this._spThemeProvider.loadThemedStyles();
            }
        }
    }
    SPPageChrome.prototype.changeSuiteNavHeight = function (pixels) {
        var top = parseInt(pixels, 10) + 'px';
        if (!this._isFlexLayoutEnabled) {
            this.appDiv.style.top = top;
            this.appDiv.style.height = "calc(100% - " + top + ")";
            this.propertyPaneContainerDiv.style.top = top;
        }
        this.suiteNavDiv.style.height = top;
    };
    Object.defineProperty(SPPageChrome.prototype, "themeProvider", {
        get: function () {
            return this._spThemeProvider;
        },
        enumerable: true,
        configurable: true
    });
    SPPageChrome.prototype.show = function () {
        if (!this._isDisposed) {
            document.body.appendChild(this._chromeDiv);
        }
    };
    SPPageChrome.prototype.hide = function () {
        if (!this._isDisposed) {
            this._chromeDiv = document.body.removeChild(this._chromeDiv);
        }
    };
    SPPageChrome.prototype.dispose = function () {
        if (_SPLoaderFlights._useNewBootSequence && !this._isDisposed) {
            delete this.suiteNavDiv;
            delete this.appDiv;
            delete this.propertyPaneContainerDiv;
            if (this.appAndPropertyPaneDiv) {
                delete this.appAndPropertyPaneDiv;
            }
            delete this.dialogContainerDiv;
            delete this._chromeDiv;
            this._isDisposed = true;
        }
    };
    Object.defineProperty(SPPageChrome.prototype, "isDisposed", {
        get: function () {
            return this._isDisposed;
        },
        enumerable: true,
        configurable: true
    });
    SPPageChrome.prototype._createPageChrome = function () {
        _PerformanceLogger.devMark('SPPageChrome.createPageChrome');
        document.body.className = 'ms-font-m';
        this._chromeDiv = document.createElement('div');
        this._chromeDiv.setAttribute('class', this._isFlexLayoutEnabled ? 'SPPageChrome' : 'SPPageChromeNoFlexbox');
        this.suiteNavDiv = document.createElement('div');
        this.suiteNavDiv.setAttribute('id', 'SuiteNavPlaceHolder');
        this.appDiv = document.createElement('div');
        this.appDiv.setAttribute('class', this._isFlexLayoutEnabled ? "SPPageChrome-app" : "SPPageChromeNoFlexbox-app " + styles.spPageContentTransition);
        this.appDiv.setAttribute('id', 'spPageChromeAppDiv');
        this.propertyPaneContainerDiv = document.createElement('div');
        if (this._isFlexLayoutEnabled) {
            this.propertyPaneContainerDiv.setAttribute('class', "spPropertyPaneContainer " + styles.spPageContentTransition);
            this.propertyPaneContainerDiv.setAttribute('id', 'spPropertyPaneContainer');
        }
        else {
            this.propertyPaneContainerDiv.setAttribute('class', "spPropertyPaneContainerNoFlexbox " + styles.spPageContentTransition);
            this.propertyPaneContainerDiv.setAttribute('id', 'spPropertyPaneContainerNoFlexbox');
        }
        this._chromeDiv.appendChild(this.suiteNavDiv);
        if (this._isFlexLayoutEnabled) {
            this.appAndPropertyPaneDiv = document.createElement('div');
            this.appAndPropertyPaneDiv.setAttribute('class', _SPFlight.isEnabled(175) ?
                'spAppAndPropertyPanelContainer spAppAndPropertyPanelNavBelowHeader' :
                'spAppAndPropertyPanelContainer');
            this._chromeDiv.appendChild(this.appAndPropertyPaneDiv);
            this.appAndPropertyPaneDiv.appendChild(this.appDiv);
            this.appAndPropertyPaneDiv.appendChild(this.propertyPaneContainerDiv);
        }
        else {
            this._chromeDiv.appendChild(this.appDiv);
            this._chromeDiv.appendChild(this.propertyPaneContainerDiv);
        }
        this.dialogContainerDiv = document.createElement('div');
        this.dialogContainerDiv.setAttribute('id', 'spDialogContainerDiv');
        this._chromeDiv.appendChild(this.dialogContainerDiv);
        if (!window.onerror) {
            window.onerror = SPGlobalErrorHandler.basicTracingHandler;
        }
    };
    Object.defineProperty(SPPageChrome.prototype, "_isFlexLayoutEnabled", {
        get: function () {
            return !_SPKillSwitch.isActivated(Guid.parse('E17AF313-E05B-4055-BDF7-7CD508F8853F'), '5/9/2017', 'Switch to flexbox');
        },
        enumerable: true,
        configurable: true
    });
    SPPageChrome.DEFER_LOAD_THEME_KILL_SWITCH = Guid.parse('3274E8D6-4E33-4E16-B223-F590B33FA789');
    return SPPageChrome;
}());
export default SPPageChrome;
