// @microsoft/node-core-library/lib/test/ProtectableMap.test.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
/// <reference types='mocha' />
const ProtectableMap_1 = require("../ProtectableMap");
const chai_1 = require("chai");
class ExampleApi {
    constructor() {
        this.clearedCount = 0;
        this.deletedCount = 0;
        this.setCount = 0;
        this._studentAgesByName = new ProtectableMap_1.ProtectableMap({
            onClear: (source) => {
                ++this.clearedCount;
            },
            onDelete: (source, key) => {
                ++this.deletedCount;
            },
            onSet: (source, key, value) => {
                ++this.setCount;
                if (key.toUpperCase() !== key) {
                    throw new Error('The key must be all upper case: ' + key);
                }
                // If the provided value is negative, clamp it to zero:
                return Math.max(value, 0);
            }
        });
    }
    get studentAgesByName() {
        return this._studentAgesByName.protectedView;
    }
    doUnprotectedOperations() {
        // These are unprotected because they interact with this._studentAgesByName
        // instead of this._studentAgesByName.protectedView.
        this._studentAgesByName.clear();
        this._studentAgesByName.set('Dave', -123);
    }
}
describe('ProtectableMap', () => {
    it('Protected operations', () => {
        const exampleApi = new ExampleApi();
        exampleApi.studentAgesByName.clear();
        exampleApi.studentAgesByName.set('ALICE', 23);
        exampleApi.studentAgesByName.set('BOB', 21);
        exampleApi.studentAgesByName.set('BOB', -1);
        exampleApi.studentAgesByName.set('CHARLIE', 22);
        exampleApi.studentAgesByName.delete('CHARLIE');
        chai_1.assert.equal(exampleApi.clearedCount, 1);
        chai_1.assert.equal(exampleApi.setCount, 4);
        chai_1.assert.equal(exampleApi.deletedCount, 1);
        chai_1.assert.equal(exampleApi.studentAgesByName.get('ALICE'), 23);
        chai_1.assert.equal(exampleApi.studentAgesByName.get('BOB'), 0); // clamped by onSet()
        chai_1.assert.equal(exampleApi.studentAgesByName.has('CHARLIE'), false);
    });
    it('Unprotected operations', () => {
        const exampleApi = new ExampleApi();
        exampleApi.doUnprotectedOperations();
        // Interacting directly with the ProtectableMap bypasses the hooks
        chai_1.assert.equal(exampleApi.clearedCount, 0);
        chai_1.assert.equal(exampleApi.studentAgesByName.get('Dave'), -123);
    });
    it('Error case', () => {
        const exampleApi = new ExampleApi();
        chai_1.assert.throw(() => {
            exampleApi.studentAgesByName.set('Jane', 23);
        }, 'The key must be all upper case: Jane');
    });
});
