// @microsoft/gulp-core-build-typescript/lib/index.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const TypeScriptTask_1 = require("./TypeScriptTask");
exports.TypeScriptTask = TypeScriptTask_1.TypeScriptTask;
const TSLintTask_1 = require("./TSLintTask");
const TextTask_1 = require("./TextTask");
const RemoveTripleSlashReferenceTask_1 = require("./RemoveTripleSlashReferenceTask");
const gulp_core_build_1 = require("@microsoft/gulp-core-build");
const ApiExtractorTask_1 = require("./ApiExtractorTask");
exports.ApiExtractorTask = ApiExtractorTask_1.ApiExtractorTask;
const TscCmdTask_1 = require("./miniTasks/TscCmdTask");
exports.TscCmdTask = TscCmdTask_1.TscCmdTask;
const TslintCmdTask_1 = require("./miniTasks/TslintCmdTask");
exports.TslintCmdTask = TslintCmdTask_1.TslintCmdTask;
const ApiExtractorStandaloneTask_1 = require("./miniTasks/ApiExtractorStandaloneTask");
__export(require("./TypeScriptConfiguration"));
exports.apiExtractor = new ApiExtractorTask_1.ApiExtractorTask();
exports.typescript = new TypeScriptTask_1.TypeScriptTask();
exports.tslint = new TSLintTask_1.TSLintTask();
exports.text = new TextTask_1.TextTask();
exports.removeTripleSlash = new RemoveTripleSlashReferenceTask_1.RemoveTripleSlashReferenceTask();
exports.tscCmd = new TscCmdTask_1.TscCmdTask();
exports.tslintCmd = new TslintCmdTask_1.TslintCmdTask();
exports.apiExtractorStandalone = new ApiExtractorStandaloneTask_1.ApiExtractorStandaloneTask();
// tslint:disable:export-name
exports.default = gulp_core_build_1.parallel(exports.tslint, gulp_core_build_1.serial(exports.typescript, exports.removeTripleSlash));
